plugins/methodinfo/src/main/java/org/qbicc/plugin/methodinfo/MethodDataEmitter.java

```diff
@@ -0,0 +1,387 @@
+                return lf.literalOf(methodInfoType, valueMap);
+        }).toArray(Literal[]::new);
+
+        return lf.literalOf(ts.getArrayType(methodInfoType, minfoLiterals.length), List.of(minfoLiterals));
+    }
+
+    Literal emitSourceCodeInfoTable(CompilationContext ctxt, SourceCodeInfo[] scList) {
+        TypeSystem ts = ctxt.getTypeSystem();
+        LiteralFactory lf = ctxt.getLiteralFactory();
+
+        CompoundType.Member[] members = new CompoundType.Member[4];
+        ValueType uint32Type = ts.getUnsignedInteger32Type();
+        int compoundTypeSize = 0;
+        for (int i = 0; i < members.length; i++) {
+            members[i] = ts.getCompoundTypeMember("f" + i, uint32Type, compoundTypeSize, uint32Type.getAlign());
+            compoundTypeSize += uint32Type.getSize();
+        }
+
+        CompoundType sourceCodeInfoType = ts.getCompoundType(CompoundType.Tag.STRUCT, "qbicc_souce_code_info", compoundTypeSize, uint32Type.getAlign(), () -> List.of(members));
+        Literal[] scInfoLiterals = Arrays.stream(scList).parallel().map(scInfo -> {
+            HashMap<CompoundType.Member, Literal> valueMap = new HashMap<>();
+            valueMap.put(sourceCodeInfoType.getMember(0), lf.literalOf(scInfo.getMethodInfoIndex()));
+            valueMap.put(sourceCodeInfoType.getMember(1), lf.literalOf(scInfo.getLineNumber()));
+            valueMap.put(sourceCodeInfoType.getMember(2), lf.literalOf(scInfo.getBcIndex()));
+            valueMap.put(sourceCodeInfoType.getMember(3), lf.literalOf(scInfo.getInlinedAtIndex()));
+            return lf.literalOf(sourceCodeInfoType, valueMap);
+        }).toArray(Literal[]::new);
+
+        return lf.literalOf(ts.getArrayType(sourceCodeInfoType, scInfoLiterals.length), List.of(scInfoLiterals));
+    }
+
+    Literal emitInstructionMap(CompilationContext ctxt, InstructionMap[] imapList) {
+        TypeSystem ts = ctxt.getTypeSystem();
+        LiteralFactory lf = ctxt.getLiteralFactory();
+        Section section = ctxt.getImplicitSection(ctxt.getDefaultTypeDefinition());
+
+        CompoundType.Member[] members = new CompoundType.Member[2];
+        int index = 0;
+        int compoundTypeSize = 0;
+        ValueType uint64Type = ts.getUnsignedInteger64Type();
+        members[0] = ts.getCompoundTypeMember("f" + (index++), uint64Type, compoundTypeSize, uint64Type.getAlign());
+        compoundTypeSize += uint64Type.getSize();
+        ValueType uint32Type = ts.getUnsignedInteger32Type();
+        members[1] = ts.getCompoundTypeMember("f" + (index++), uint32Type, compoundTypeSize, uint64Type.getAlign());
+        compoundTypeSize += uint32Type.getSize();
+
+        CompoundType imapType = ts.getCompoundType(CompoundType.Tag.STRUCT, "qbicc_instruction_map", compoundTypeSize, uint64Type.getAlign(), () -> List.of(members));
+
+        Literal[] imapLiterals = IntStream.range(0, imapList.length)
+            .parallel()
```

Comment:
Will these be large enough to benefit from the use of parallel streams?  A rough estimate is that there should be at least 10_000 items for the cost of splitting across threads to be worth it.

Reply:
Probably not when emitting string tables for file names and class names, but I am guessing method name and method desc string tables can scale to 10_000 values depending on app size. For our sample program I get around 2500 entries in method name and desc tables.